### ISSUE TITLE
Exceptions crashing daq_application

### ISSUE THREAD

**Issue by jcfreeman2** [DUNE-DAQ/appfwk#133]
More precisely: exceptions which are thrown from functions wrapped by `appfwk::ThreadHelper` will end `daq_application`. If I understand correctly, we don't want user code to end `daq_application`s but want them to flag an error which Run Control can then handle. 

Seeing this in action is simple. E.g., you can build `listrev` as described in https://github.com/DUNE-DAQ/appfwk/wiki/Compiling-and-running-under-v2.0.0 but before running `list-reversal-app.json` set the number of ints produced by a vector to a negative number, e.g., 
```
sed -r -i 's/"nIntsPerList": 5/"nIntsPerList": -1/' sourcecode/listrev/test/list-reversal-app.json 
daq_application -c stdin://sourcecode/listrev/test/list-reversal-app.json
```
or for that matter you can do something similar with `appfwk/schema/fdpc-job.json`. In both cases an `std::bad_alloc` will sail out of `daq_application`, ending it. 

We should think of ways to make it as difficult as possible for DAQ module developers to crash `daq_application` with their code; it would be worth investigating what the best way to keep an `appfwk::ThreadHelper`'s function from doing so would be.

**Comment by philiprodrigues** [comment]
Some possibilities:

* As in #145, we could wrap the user function with code that stores any exceptions in a `std::exception_ptr`. The `exception_ptr` can be used to re-throw the exception in `stop_working_thread()`
* A more flexible approach, suggested by Marco, would be to use `std::async` and get a `std::future<void>` back. We could pass the future to the caller of `start_working_thread()`, or have it available via some `get_future()` function. Caller then has the option to wait on the future, or call `future.get()` to have the exception re-thrown wherever caller desires. (I don't see a way to explicitly check whether a future is ready, but `wait_for` with timeout=0 would presumably achieve that). If the future has not been passed out of `ThreadHelper` via `get_future()`, `stop_working_thread()` could (optionally) rethrow the exception
* We could use either of the above techniques to prevent the exception crashing the entire application, but just never rethrow it?

**Comment by mroda88** [comment]
Regardless of what is the way we use to retrieve the exception, in order to decide if rethrow or not, I guess we want to decide what we want to achieve with the rethrowing. Considering that the rethrowing will be triggered by the stop call, the exception will probably be very late to be useful. 
But I can see a reason to rethrow. We could use it that the `stop` could not happen as expected as the thread died. Probably this means that the queues were not drained and so the right thing to do, after attempting the `stop` is to notify the control that a `scrap` rather than call a new `start` is the way forward. Because of that I'm leaning toward a rethrow at the moment, but I reckon it's not much.
Any other idea on what we could do with a late exception?

**Comment by alessandrothea** [comment]
Closed by #30
